### ISSUE TITLE
Quaternion getEulerAnglesXYZ NaN

### ISSUE THREAD

**Issue by haubna** [JOML-CI/JOML#217]
Due to the nature of floating point precision numbers
`eulerAngles.y = Math.asin(2.0 * (x*z + y*w));`
results sometimes in NaN because this `2.0 * (x*z + y*w)` is something like `1.000000002` (even after normalizing the quaternion). Is this an expected behaviour? Seems odd to me to return NaN because of an precision error.

**Comment by haubna** [comment]
I am using this (implementation is not by me, I have no clue about quaternions)
```
double test = quat.x * quat.z + quat.y * quat.w;

if (test > 0.499998) { // singularity at north pole
	dst.x = 2 * (double) java.lang.Math.atan2(quat.x, quat.w);
	dst.y = (double) java.lang.Math.PI / 2;
	dst.z = 0;
	return dst;
}

if (test < -0.499998) { // singularity at south pole
	dst.x = -2 * (double) java.lang.Math.atan2(quat.x, quat.w);
	dst.y = (double) -java.lang.Math.PI / 2;
	dst.z = 0;
	return dst;
}

dst.x = Math.atan2(2.0 * (quat.x*quat.w - quat.y*quat.z), 1.0 - 2.0 * (quat.x*quat.x + quat.y*quat.y));
dst.y = Math.asin(2.0 * test);
dst.z = Math.atan2(2.0 * (quat.z*quat.w - quat.x*quat.y), 1.0 - 2.0 * (quat.y*quat.y + quat.z*quat.z));
```
instead of this
```
eulerAngles.x = Math.atan2(2.0 * (x*w - y*z), 1.0 - 2.0 * (x*x + y*y));
eulerAngles.y = Math.asin(2.0 * (x*z + y*w));
eulerAngles.z = Math.atan2(2.0 * (z*w - x*y), 1.0 - 2.0 * (y*y + z*z));
return eulerAngles;
```
Works like a charm for me. But testing for 0.4999998 is still a pretty ugly solution.

**Comment by httpdigest** [comment]
Fix is published as new 1.9.23-SNAPSHOT.
